Zero-Copy ReadSlice Trait

krakow10 opened the pull request:
The point of these changes is to reduce the number of copies where possible. To achieve this, we split off chunks of a byte slice instead of allocating a String or Vec<u8> where applicable.

### Performance
5-8% faster on the deserialize microbenchmarks according to my testing.  This mostly affects performance on a per-prop-chunk basis rather than per-property which explains why it doesn't seem to have a big impact on the Miner's Haven benchmark.  This is probably also why #545 didn't see a big performance uplift.

krakow10 committed:
rbx_binary: zero-copy slice readers

krakow10 committed:
rbx_binary: use zero copy string for prop_name

krakow10 committed:
rbx_binary: use zero copy for applicable binary strings

krakow10 committed:
rbx_binary: use zero copy string for type_name

Dekkonot commented on rbx_binary/src/core.rs:
```diff
@@ -315,6 +315,48 @@ pub trait RbxWriteExt: Write {
+    let length = slice.read_le_u32()?;
+
+    let out;
+    // split_at can panic if the slice is shorter than length
+    // but we do not expect that to happen.
```
Do you think it's worth checking? We could produce a meaningful error if this invariant was violated, which might help consumers.

Then again, well-made producers of rbxm files will never have this problem so it may be a moot point; if someone is hand-crafting binary files they should be careful.

krakow10 replied:
I don't think it's worth it.  Seems like the kind of thing that would be quickly found and fixed in any encoder implementation.  It can always be added later if the situation calls for it, like perhaps if some legacy tool that will never be fixed is creating files malformed in this specific way.  It'd be simple to call `.split_at_checked` and `.ok_or` a `UnexpectedEof` io error, but not sure how that would affect performance (probably negligable).  Let me know if you want it.

Dekkonot commented on rbx_binary/src/core.rs:
```diff
@@ -315,6 +315,48 @@ pub trait RbxWriteExt: Write {
+    // but we do not expect that to happen.
+    (out, *slice) = slice.split_at(length as usize);
+
+    let Ok(str) = core::str::from_utf8(out) else {
+        return Err(io::Error::new(
+            io::ErrorKind::InvalidData,
+            "stream did not contain valid UTF-8",
+        ));
+    };
+
+    Ok(str)
```
```suggestion
    core::str::from_utf8(out).map_err(|_| {
        io::Error::new(
            io::ErrorKind::InvalidData,
            "stream did not contain valid UTF-8",
        )
    })
```

No need for the explicit return here; we can just use `map_err`.

Dekkonot reviewed (changes requested):
I'm a clever boy and can work out that the new functions aren't a part of `RbxReadExt` because they operate explicitly on a slice instead of a nebulous `impl Read`, but it might be worth writing it down for future generations. I'll leave that to your discretion, but I tend to err on the side of writing things down even if they feel obvious because it lowers the bus factor for the project.

Otherwise this seems fine to me.

krakow10 committed:
Update rbx_binary/src/core.rs

Co-authored-by: Micah <dekkonot@rocketmail.com>

krakow10 committed:
note that functions are not part of RbxReadExt

krakow10 commented:
> `RbxReadExt` ... might be worth writing it down

Done.  I was considering a trait too, but decided this was simpler.

krakow10 committed:
replace duplicate code with function call

krakow10 committed:
fix rebase

krakow10 committed:
construct interleaved iterators directly using slice

krakow10 committed:
add comments about function origin

krakow10 committed:
fix doc comments

krakow10 committed:
infallible signature

the split_at can panic, there is a checked variant if it should be handled gracefully

krakow10 committed:
Revert "infallible signature"

This reverts commit 030f68a2a0cd93572cf87313c4685d477b361fb5.

krakow10 committed:
check slice len

krakow10 committed:
check slice len

krakow10 committed:
remove it variable

krakow10 committed:
remove read_binary_string & read_string from RbxReadExt

Utilize ToOwned to copy strings when necessary.

krakow10 committed:
fix doc

krakow10 committed:
change to trait

krakow10 commented:
This now combines changes from #561 to create a RbxReadZeroCopy trait.

krakow10 committed:
restore original test name

krakow10 committed:
add cold path like std Vec does

krakow10 committed:
unify split_at_checked as one common base function

krakow10 committed:
move trait bound down

krakow10 committed:
take advantage of read_slice

krakow10 committed:
take advantage of read_slice in text_deserializer

krakow10 committed:
appease clippy

krakow10 committed:
put it back into RbxReadExt, but with trait bounds

krakow10 committed:
improve read_slice doc

krakow10 committed:
RbxReadInterleaved trait for grouping and loc

krakow10 committed:
move RbxReadZeroCopy code to a more localized place

krakow10 committed:
rename RbxReadZeroCopy to ReadSlice

krakow10 committed:
improve doc wording

krakow10 committed:
simplify remaining data in text_deserializer

krakow10 committed:
use read_slice for SharedScript hash

Dekkonot commented:
> I would like to open stacked PRs, but they need to be created with a branch within the repository itself, which I don't have access to.

If you are too keen, you can make PRs on your own fork of rbx-dom and link them. I'd probably just merge them with a squash though so you can theoretically rebase after they're merged.

krakow10 committed:
remove generic from test

krakow10 committed:
Revert "remove generic from test"

This reverts commit 6b5ac80b1ab554b799e6eba7dcca3257e6c6ec10.

krakow10 committed:
Merge branch 'master' into read-slice_zero-copy

krakow10 committed:
split RbxReadInterleaved into separate PR

krakow10 committed:
Merge branch 'read-slice_zero-copy2' into read-slice_zero-copy

krakow10 commented:
I've split out the RbxReadInterleaved change into its own PR.  This is just the ReadSlice trait now.

krakow10 committed:
use ? instead of unwrap

krakow10 committed:
Merge branch 'read-slice_zero-copy2' into read-slice_zero-copy

kennethloeffler reviewed (dismissed):
Nice, this looks to be a significant optimization for light workloads: I saw about +10% on the small benchmarks. It makes sense, since we have proportionally more strings to copy in larger files. Code looks great to me, nice work!

krakow10 commented:
It's good to hear that I'm not just on a wild goose chase trying to find improvements to this PR, that the end result is actually desirable.  Thanks for the feedback.

krakow10 committed:
Merge branch 'master' into read-slice_zero-copy

krakow10 committed:
fix clippy lint

krakow10 committed:
format

krakow10 committed:
update submodule

krakow10 commented:
Fixed merge conflicts.

krakow10 committed:
Merge branch 'master' into read-slice_zero-copy

krakow10 committed:
Merge branch 'master' into read-slice_zero-copy

krakow10 committed:
read optional markers as a slice

krakow10 committed:
Merge branch 'master' into read-slice_zero-copy
